netaddr_test.go

```diff
@@ -1441,31 +1445,38 @@ func BenchmarkIPv4Contains(b *testing.B) {
 	}
 }
 
-func BenchmarkParseIPv4(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		ParseIP("192.168.1.1")
-	}
-}
-
-func BenchmarkParseIPv6(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		ParseIP("fe80::1cc0:3e8c:119f:c2e1%ens18")
-	}
+var parseBenchInputs = []struct {
+	name string
+	ip   string
+}{
+	{"v4", "192.168.1.1"},
+	{"v6", "fd7a:115c:a1e0:ab12:4843:cd96:626b:430b"},
+	{"v6_ellipsis", "fd7a:115c::626b:430b"},
+	{"v6_v4", "::ffff:192.168.140.255"},
+	{"v6_zone", "1:2::ffff:192.168.140.255%eth1"},
 }
 
-func BenchmarkStdParseIPv4(b *testing.B) {
-	b.ReportAllocs()
-	for i := 0; i < b.N; i++ {
-		net.ParseIP("192.168.1.1")
+func BenchmarkParseIP(b *testing.B) {
+	z := intern.Get("eth1") // Pin to not benchmark the intern package
+	for _, test := range parseBenchInputs {
+		b.Run(test.name, func(b *testing.B) {
+			b.ReportAllocs()
+			for i := 0; i < b.N; i++ {
+				sinkIP, _ = ParseIP(test.ip)
+			}
+		})
 	}
+	_ = z // keep the interned value alive during sub-benchmarks
```

Comment:
This will get optimized away. Use a global sink instead. That might explain the allocation. If it doesn't explain the allocation, run with `-memprofile` to find it. (I can help with that if needed.)

Reply:
No change. memprofile says the 16B allocation happens in `IP.WithZone`, at `ip.z = intern.Get(zone)`. That the allocation is 16B makes me suspect the conversion of zone from string to interface{}. I'm not clear on why that allocation needs to happen.

Reply:
With gcflags `-m -m`:

```
./netaddr.go:758:42: parameter zone leaks to {heap} with derefs=0:
./netaddr.go:758:42:   flow: {heap} = zone:
./netaddr.go:758:42:     from ip.WithZone(zone) (call parameter) at ./netaddr.go:769:19
```

Reply:
Loose hypothesis: in creating the interface{} for the call to intern.Get, the compiler can't prove that the interface won't outlive the zone string, and so is forced to evict it to the heap.

In practice, if there's a cache hit within intern.Get, the interface{} won't outlive the WithZone call, but the compiler can't prove it, and so any WithZone call with a non-empty zone string will force a heap allocation for a string slice.

I don't think there's any way to avoid that, unless intern exposes a specialized GetString API (would be easier with generics :( ).

Reply:
Also checked with -cpuprofile, and confirmed that v6+zone's slowness comes entirely from WithZone, about 50% spent allocating that escaping string, and the other 50% in intern.Get doing the map lookup and whatnot.

So, the conclusion is just: at the timescales we're working at here, using the intern package is quite expensive, and unless we can make it significantly cheaper, v6+zone will always be ~2x slower than the stdlib. My take is that the slowness for that edge case is worth it in return for the other nice things that interned zoneptrs get us.